File: ios/HybridRiveFontConfig.swift
import CoreText
import NitroModules
import RiveRuntime

class HybridRiveFontConfig: HybridRiveFontConfigSpec {
  private static let defaultWeight = 0
  private static var fontsByWeight: [Int: [any HybridFallbackFontSpec]] = [:]

  func loadFontFromURL(url: String) throws -> Promise<(any HybridFallbackFontSpec)> {
    return Promise.async {
      guard let parsedURL = URL(string: url) else {
        throw RuntimeError.error(withMessage: "Invalid font URL: \(url)")
      }
      let (data, _) = try await URLSession.shared.data(from: parsedURL)
      let font = try Self.createUIFont(from: data)
      return HybridFallbackFont(font: font)
    }
  }

  func loadFontFromResource(resource: String) throws -> (any HybridFallbackFontSpec) {
    let nsResource = resource as NSString
    let name = nsResource.deletingPathExtension
    let ext = nsResource.pathExtension.isEmpty ? nil : nsResource.pathExtension

    guard let path = Bundle.main.path(forResource: name, ofType: ext) else {
      throw RuntimeError.error(withMessage: "Font resource not found: \(resource)")
    }
    let data = try Data(contentsOf: URL(fileURLWithPath: path))
    let font = try Self.createUIFont(from: data)
    return HybridFallbackFont(font: font)
  }

  func loadFontFromBytes(bytes: ArrayBuffer) throws -> (any HybridFallbackFontSpec) {
    let data = bytes.toData(copyIfNeeded: true)
    let font = try Self.createUIFont(from: data)
    return HybridFallbackFont(font: font)
  }

  func loadFontByName(name: String) throws -> (any HybridFallbackFontSpec) {
    guard let font = UIFont(name: name, size: UIFont.systemFontSize) else {
      throw RuntimeError.error(withMessage: "System font not found: \(name)")
    }
    return HybridFallbackFont(font: font)
  }

  func getSystemDefaultFont() throws -> (any HybridFallbackFontSpec) {
    return HybridDefaultFallbackFont()
  }

  func setFontsForWeight(weight: Double, fonts: [(any HybridFallbackFontSpec)]) throws {
    let key = Int(weight)
    Self.fontsByWeight[key] = fonts
  }

  func applyFallbackFonts() throws -> Promise<Void> {
    return Promise.async {
      _ = RiveFont.self
      RiveFont.fallbackFontsCallback = { weight in
        let requestedWeight = Int(weight.rawWeight)
        let specs = Self.fontsByWeight[requestedWeight] ?? Self.fontsByWeight[Self.defaultWeight] ?? []
        return specs.compactMap { spec in
          if spec is HybridDefaultFallbackFont {
            return RiveFallbackFontDescriptor()
          } else if let fallbackFont = spec as? HybridFallbackFont {
            return fallbackFont.font
          } else {
            print("[RiveFonts] Unknown fallback font spec type: \(type(of: spec))")
            return nil
          }
        }
      }
    }
  }

  func clearFallbackFonts() throws -> Promise<Void> {
    return Promise.async {
      Self.fontsByWeight.removeAll()
      RiveFont.fallbackFontsCallback = { _ in [RiveFallbackFontDescriptor()] }
      RiveFont.fallbackFonts = [RiveFallbackFontDescriptor()]
    }
  }

  private static func createUIFont(from data: Data) throws -> UIFont {
    guard let provider = CGDataProvider(data: data as CFData),
      let cgFont = CGFont(provider)
    else {
      throw RuntimeError.error(withMessage: "Failed to decode font data")
    }

    var error: Unmanaged<CFError>?
    if !CTFontManagerRegisterGraphicsFont(cgFont, &error) {
      let cfError = error?.takeRetainedValue()
      let domain = cfError.map { CFErrorGetDomain($0) as String } ?? ""
      if domain != kCTFontManagerErrorDomain as String {
        throw RuntimeError.error(
          withMessage: "Failed to register font: \(cfError?.localizedDescription ?? "unknown error")"
        )
      }
    }

    guard let fontName = cgFont.postScriptName as String? else {
      throw RuntimeError.error(withMessage: "Failed to get font name from data")
    }
    guard let font = UIFont(name: fontName, size: UIFont.systemFontSize) else {
      throw RuntimeError.error(withMessage: "Failed to create UIFont for: \(fontName)")
    }
    return font
  }
}

File: android/src/main/java/com/margelo/nitro/rive/HybridRiveFontConfig.kt
package com.margelo.nitro.rive

import android.util.Log
import androidx.annotation.Keep
import app.rive.runtime.kotlin.fonts.FontFallbackStrategy
import app.rive.runtime.kotlin.fonts.FontHelper
import app.rive.runtime.kotlin.fonts.Fonts
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import java.net.URL
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

@Keep
@DoNotStrip
class HybridRiveFontConfig : HybridRiveFontConfigSpec() {
  companion object {
    private const val TAG = "RiveFonts"
    private const val DEFAULT_WEIGHT = 0
    private val fontsByWeight: MutableMap<Int, List<HybridFallbackFontSpec>> =
      java.util.Collections.synchronizedMap(mutableMapOf())

    private fun resetFontCache() {
      try {
        FontFallbackStrategy.cppResetFontCache()
      } catch (e: Exception) {
        Log.e(TAG, "Failed to reset font cache: ${e.message}")
      }
    }
  }

  override fun loadFontFromURL(url: String): Promise<HybridFallbackFontSpec> {
    return Promise.async {
      val fontBytes = withContext(Dispatchers.IO) {
        URL(url).readBytes()
      }
      HybridFallbackFont(fontBytes)
    }
  }

  override fun loadFontFromResource(resource: String): HybridFallbackFontSpec {
    val context = NitroModules.applicationContext
      ?: throw Error("Application context not available")

    val resName = resource.substringBeforeLast(".")

    val rawId = context.resources.getIdentifier(resName, "raw", context.packageName)
    if (rawId != 0) {
      val fontBytes = context.resources.openRawResource(rawId).use { it.readBytes() }
      return HybridFallbackFont(fontBytes)
    }

    val fontId = context.resources.getIdentifier(resName, "font", context.packageName)
    if (fontId != 0) {
      val fontBytes = context.resources.openRawResource(fontId).use { it.readBytes() }
      return HybridFallbackFont(fontBytes)
    }

    val assetPaths = listOf("fonts/$resource", resource)
    for (assetPath in assetPaths) {
      try {
        val fontBytes = context.assets.open(assetPath).use { it.readBytes() }
        return HybridFallbackFont(fontBytes)
      } catch (_: Exception) {
      }
    }

    throw Error("Font resource not found: $resource (checked res/raw, res/font, assets/fonts)")
  }

  override fun loadFontFromBytes(bytes: ArrayBuffer): HybridFallbackFontSpec {
    val buffer = bytes.getBuffer(false)
    val byteArray = ByteArray(buffer.remaining())
    buffer.get(byteArray)
    return HybridFallbackFont(byteArray)
  }

  override fun loadFontByName(name: String): HybridFallbackFontSpec {
    val fonts = FontHelper.getFallbackFonts(Fonts.FontOpts(familyName = name))
    if (fonts.isEmpty()) {
      throw Error("System font not found: $name")
    }
    val fontBytes = FontHelper.getFontBytes(fonts.first())
      ?: throw Error("Could not read font bytes for: $name")
    return HybridFallbackFont(fontBytes)
  }

  override fun getSystemDefaultFont(): HybridFallbackFontSpec {
    return HybridDefaultFallbackFont()
  }

  override fun setFontsForWeight(weight: Double, fonts: Array<HybridFallbackFontSpec>) {
    val key = weight.toInt()
    synchronized(fontsByWeight) {
      fontsByWeight[key] = fonts.toList()
    }
  }

  override fun applyFallbackFonts(): Promise<Unit> {
    return Promise.async {
      FontFallbackStrategy.stylePicker = object : FontFallbackStrategy {
        override fun getFont(weight: Fonts.Weight): List<ByteArray> {
          val requestedWeight = weight.weight
          val specs = synchronized(fontsByWeight) {
            fontsByWeight[requestedWeight] ?: fontsByWeight[DEFAULT_WEIGHT] ?: emptyList()
          }
          return specs.mapNotNull { spec ->
            when (spec) {
              is HybridDefaultFallbackFont ->
                FontHelper.getFallbackFontBytes(Fonts.FontOpts(weight = weight))
              is HybridFallbackFont ->
                spec.fontBytes
              else -> {
                Log.e(TAG, "Unknown fallback font spec type: ${spec::class.simpleName}")
                null
              }
            }
          }
        }
      }
      resetFontCache()
    }
  }

  override fun clearFallbackFonts(): Promise<Unit> {
    return Promise.async {
      synchronized(fontsByWeight) {
        fontsByWeight.clear()
      }
      FontFallbackStrategy.stylePicker = null
      resetFontCache()
    }
  }
}
